power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/observability.hpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "measured_values.hpp"
+
+#include "../common/exception.hpp"
+
+namespace power_grid_model::math_solver {
+
+template <symmetry_tag sym>
+inline void necessary_observability_check(MeasuredValues<sym> const& measured_values,
+                                          std::shared_ptr<MathModelTopology const> const& topo) {
+
+    Idx const n_bus{topo->n_bus()};
+    std::vector<BranchIdx> const& branch_bus_idx{topo->branch_bus_idx};
+    Idx n_voltage_magnitude_sensor{};
+    Idx n_voltage_phasor_sensor{};
+    Idx n_branch_sensor{};
+    Idx n_injection_sensor{};
+
+    // count voltage measurements
+    for (Idx bus = 0; bus != n_bus; ++bus) {
+        if (measured_values.has_voltage(bus)) {
+            n_voltage_magnitude_sensor++;
+            if (measured_values.has_angle_measurement(bus)) {
+                n_voltage_phasor_sensor++;
+            }
+        }
+    }
+
+    if (n_voltage_magnitude_sensor + n_voltage_phasor_sensor < 1) {
+        throw NotObservableError{};
```

Comment:
should we provide an additional message here? e.g. `At least one voltage measurement is required`?